### PR TITLE
Collection expressions: prefer ReadOnlySpan<T> over Span<T> in overload resolution

### LINKED ISSUE #70318: Overload resolution prefers Span<T> over ReadOnlySpan<T> for collection expression
**Version Used**: 

**Steps to Reproduce**:

Compile and run:
```csharp
using System;

class Program
{
    static void Main()
    {
        M([1, 2, 3]);
    }
    
    static void M(Span<int> span) { Console.WriteLine("Span<int>"); }
    static void M(ReadOnlySpan<int> span) { Console.WriteLine("ReadOnlySpan<int>"); }
}
```

See [sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK7wCYgNQB8ACATAIwCwAUBUQAT7EDsFA3hdW7cQGy0As1AstBgAKAJSt2LcuxkDhAbWJxqhZQGYAuqIDcEtgF891I3W74+/YQGUADhBgAeWAgB81AM52Yo6kw4BOYQAiW3snGFcgnWpDaXZTXjkAJQBTCAwAeRgAGwBPUMdnN097Hz86QKDU9Ky8gvDI6Nj9IA===)

**Expected Behavior**:
`ReadOnlySpan<int>`

**Actual Behavior**:
`Span<int>`

Reported by @stephentoub

### PR DESCRIPTION
**Pull request by cston**
See updated *better conversion from expression*: https://github.com/dotnet/csharplang/pull/7591

Fixes #70318

### CODE REVIEW

**Review by 333fred — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 1]
```diff
@@ -2531,57 +2531,87 @@ private BetterResult BetterConversionFromExpression(BoundExpression node, TypeSy
             if (!conv2.IsConditionalExpression && conv1.IsConditionalExpression)
                 return BetterResult.Right;
 
-            // C1 and C2 are collection expression conversions and
-            // T1 is a span_type with iteration type E1, and
-            // T2 is an array_or_array_interface_or_string_type with iteration type E2, and
-            // - E1 is implicitly convertible to E2
+            // - E is a collection expression and one of the following holds:
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:2534]:
```suggestion
            // - E is a collection expression and one of the following holds: ...
```

[the remark this answers is not in the document]
Comment (cston) [reply src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs]: Added "..." to indicate there's more, without copying the entire content here.

File: src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs [hunk 2]
```diff
@@ -2531,57 +2531,87 @@ private BetterResult BetterConversionFromExpression(BoundExpression node, TypeSy
...
             }
 
             // - T1 is a better conversion target than T2 and either C1 and C2 are both conditional expression
             //   conversions or neither is a conditional expression conversion.
             return BetterConversionTarget(node, t1, conv1, t2, conv2, ref useSiteInfo, out okToDowngradeToNeither);
         }
 
-        private bool IsSZArrayOrArrayInterfaceOrString(TypeSymbol type, out TypeWithAnnotations elementType)
+        // Implements the rules for
+        // - E is a collection expression and one of the following holds:
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolution.cs:2555]:
```suggestion
        // - E is a collection expression and one of the following holds: ...
```

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 3]
```diff
@@ -1616,6 +1726,75 @@ static void Main()
...
+        }
+
+        [Fact]
+        public void BetterConversionFromExpression_08A()
+        {
+            string source = """
+                class Program
+                {
+                    static void F1(int[] value) { }
+                    static void F1(object[] value) { }
```
Comment (alrz) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:1766]: I don't remember the exact notes but isn't this supposed to be unambiguous just like (int,int,int) and (object,object,object)?

Comment (cston) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs]: I'm not aware of a decision to prefer one array type over another for a collection expression conversion in overload resolution, other than the [existing rule](https://github.com/dotnet/csharpstandard/blob/standard-v6/standard/expressions.md#11646-better-conversion-target) to prefer type `T1` over `T2` if there is an implicit conversion from `T1` to `T2`.

Comment (alrz) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs]: My mistake, I thought most-specific rules can interact with the common-type from elements.

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs [hunk 4]
```diff
@@ -1616,6 +1752,75 @@ static void Main()
...
+            var comp = CreateCompilation(
+                source,
+                targetFramework: TargetFramework.Net80);
+            comp.VerifyEmitDiagnostics(
+                // (10,9): error CS0121: The call is ambiguous between the following methods or properties: 'Program.F1(ReadOnlySpan<int>)' and 'Program.F1(ReadOnlySpan<object>)'
+                //         F1([1, 2, 3]);
+                Diagnostic(ErrorCode.ERR_AmbigCall, "F1").WithArguments("Program.F1(System.ReadOnlySpan<int>)", "Program.F1(System.ReadOnlySpan<object>)").WithLocation(10, 9),
+                // (11,9): error CS0121: The call is ambiguous between the following methods or properties: 'Program.F2(Span<string>)' and 'Program.F2(Span<object>)'
+                //         F2(["a", "b"]);
+                Diagnostic(ErrorCode.ERR_AmbigCall, "F2").WithArguments("Program.F2(System.Span<string>)", "Program.F2(System.Span<object>)").WithLocation(11, 9));
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs:1782]: This doesn't feel right to me. I think we should resolve at the same time as #70238

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs (hunk 3 above)
Comment (alrz) [reply src/Compilers/CSharp/Test/Emit2/Semantics/CollectionExpressionTests.cs]: Relevant notes: https://github.com/dotnet/csharplang/blob/main/meetings/2023/LDM-2023-09-20.md
Though that's only a special-case around Span and array to resolve this, and only in one direction.

**Pull request merged**